Modules/Microsoft.PowerShell.UnixCompleters/Microsoft.PowerShell.UnixCompleters/UtilCompleterInitializer.cs

```diff
@@ -0,0 +1,143 @@
+                {
+                    object executionContext = s_executionContextProperty.GetValue(Runspace.DefaultRunspace);
+                    
+                    var completerTable = (Dictionary<string, ScriptBlock>)s_nativeArgumentCompletersProperty.GetValue(executionContext);
+
+                    if (completerTable == null)
+                    {
+                        completerTable = new Dictionary<string, ScriptBlock>(StringComparer.OrdinalIgnoreCase);
+                        s_nativeArgumentCompletersProperty.SetValue(executionContext, completerTable);
+                    }
+
+                    s_nativeArgumentCompleterTable = completerTable;
+                }
+
+                return s_nativeArgumentCompleterTable;
+            }
+        }
+
+        internal static IUnixUtilCompleter UnixUtilCompleter { get; set; }
+    }
+
+    public class UtilCompleterInitializer : IModuleAssemblyInitializer
+    {
+        private const string SHELL_PREFERENCE_VARNAME = "COMPLETION_SHELL_PREFERENCE";
+
+        public void OnImport()
+        {
+            string preferredCompletionShell = Environment.GetEnvironmentVariable(SHELL_PREFERENCE_VARNAME);
+
+            ShellType shellType = ShellType.None;
+            string shellExePath;
+            if ((string.IsNullOrEmpty(preferredCompletionShell) || !UnixHelpers.TryFindShell(preferredCompletionShell, out shellExePath, out shellType))
+                && !UnixHelpers.TryFindFallbackShell(out shellExePath, out shellType))
+            {
+                WriteError("Unable to find shell to provide unix utility completions");
+                return;
+            }
+
+            IUnixUtilCompleter utilCompleter;
+            switch (shellType)
+            {
+                case ShellType.Bash:
+                    utilCompleter = new BashUtilCompleter(shellExePath);
+                    break;
+
+                case ShellType.Zsh:
+                    utilCompleter = new ZshUtilCompleter(shellExePath);
+                    break;
+
+                default:
+                    WriteError("Unable to find shell to provide unix utility completions");
+                    return;
+            }
+
+            IEnumerable<string> utilsToComplete = utilCompleter.FindCompletableCommands();
+
+            CompleterGlobals.CompletedCommands = utilsToComplete;
+            CompleterGlobals.UnixUtilCompleter = utilCompleter;
+
+            RegisterCompletersForCommands(utilsToComplete);
+        }
+
+        private void RegisterCompletersForCommands(IEnumerable<string> commands)
+        {
+            foreach (string command in commands)
+            {
+                CompleterGlobals.NativeArgumentCompleterTable[command] = UnixUtilCompletion.CreateInvocationScriptBlock(command);
+            }
+        }
+
+        private void WriteError(string errorMessage)
+        {
+            using (var pwsh = SMA.PowerShell.Create())
```

Comment:
Looks expensive. Can we use current context?

Reply:
In fact this method doesn't work at all, since the runspace the user is looking at is occupied. So this method does nothing.

I wouldn't mind expensive error reporting, since we only do it in an exception case, but right now this method is quietly broken.

The problem is that fixing it is hard, since we need to somehow report an error from an `IModuleAssemblyInitializer`. We could throw, but the logic of import module isn't set up to catch that I don't think... That leaves trying to set up some kind of event to warn the user? I'm not sure. This essentially represents a flaw in PowerShell's module initialization logic to me.

Reply:
If it is at module import time I'd throw.

Reply:
The problem with throwing is that the module import code doesn't seem to handle an exception here, so I'm not sure if we'd cause some bad state by doing this